Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,415 @@
+
+Character classes in other languages match at either the Unicode scalar value level, or even the code unit level, instead of recognizing grapheme clusters as characters. When matching the `.` character class, other languages will only match the first part of an `"e\u{301}"` grapheme cluster. Some languages, like Perl, Ruby, and Java, support an additional `\X` metacharacter, which explicitly represents a single grapheme cluster.
+
+| Matching  `"Cafe\u{301}"` | Pattern: `^Caf.` | Remaining | Pattern:  `^Caf\X` | Remaining |
+|---|---|---|---|---|
+| C#, Rust, Go | `"Cafe"` | `"´"` | n/a | n/a |
+| NSString, Java, Ruby, Perl | `"Cafe"` | `"´"` | `"Café"` | `""` |
+
+Other than Java's `CANON_EQ` option, the vast majority of other languages and engines are not capable of comparing with canonical equivalence.
+
+</details>
+
+[SE-0211 Unicode Scalar Properties][scalarprops] added basic building blocks for classification of scalars by surfacing Unicode data from the [UCD][ucd]. [SE-0221: Character Properties][charprops] defined grapheme-cluster semantics for Swift for a subset of these. But, many classifications used in string processing are combinations of scalar properties or ad-hoc listings, and as such are not present today in Swift.
+
+Regardless of any syntax or underlying formalism, classifying characters is a worthy and much needed addition to the Swift standard library. We believe our thorough treatment of every character class found across many popular regex engines gives Swift a solid semantic basis.
+
+## Proposed Solution
+
+This pitch is narrowly scoped to Swift definitions of character classes found in regexes. For each character class, we propose:
+
+- A name for use in API
+- A `Character` API, by extending Unicode scalar definitions to grapheme clusters
+- A `Unicode.Scalar` API with modern Unicode definitions
+- If applicable, a `Unicode.Scalar` API for notable standards like POSIX (or JS?)
+
+We're proposing what we believe to be the Swiftiest definitions, referencing Unicode's [UTS\#18][uts18], [PCRE][pcre], [perl][perl], [Raku][raku], [Rust][rust], [Python][python], [C\#][csharp], [`NSRegularExpression` / ICU][icu], [POSIX][posix], [Oniguruma][oniguruma], (grep?), [Go][go], [C++][cplusplus], [RE2][re2], [Java][java], and [ECMAScript][ecmascript].
+
+To extend scalar semantics to grapheme clusters, we're using algebra and rationale from [SE-0221: Character Properties][charpropsrationale].
+
+## Detailed Design
+
+In general, we choose the Unicode recommendations in [UTS\#18][uts18] as the basis for matching character classes to `Unicode.Scalar`s. This is similar to the behavior of several languages' regular expressions when in Unicode mode (such as Perl and Python), as well as ICU-based regular expressions like `NSRegularExpression`.
+
+To extend this matching behavior to `Character`s, we treat a `Character` as belonging to a character class if its leading Unicode scalar belongs to that class.
```

Comment:
This is very much a case-by-case decision and I don't think it makes sense to declare a particular rationale unilaterally. Even the existing Character properties, restricted as they were to what was easier to express and less controversial, had two different ways.